### DISCUSSION TITLE
Feature Request: constraint_validations to respect default values

### DISCUSSION THREAD

**Discussion by 3ynm** [jeremyevans/sequel#2023] [General]
In order to have a single source of truth it would be neat that the constraint_validations extension respected the default values at the time of initialization (or otherwise, validation) of an object.

Currently a validation error is thrown if one does not initialize the default values already set by the database, and the attribute is required.

**Comment by jeremyevans** [comment]
Currently, you should use `allow_nil: true` for constraint validations on columns with default values, since in the case where a value isn't provided, you can assume the default value is valid.  I think the only issue in that case is when an explicit `nil` value is provided.  If you are using auto_validations, then auto_validations should handle that case.  Do you think that would work for your use case?

In terms of actually fixing this, the constraint_validations metadata table doesn't currently support the `allow_missing` option, and adding it would make backwards compatibility very tricky.  However, maybe the constraint_validations plugin could check the schema just as auto_validations does, determine which columns do not allow NULL and have default values, and add the `:allow_missing` option to constraint validations for those columns.  I'll see if that approach will work.

**Reply by 3ynm** [reply]
I could allow `nil` values and that would work for my current use case. But it's possible that in the future some `nil` values end up on the field, for instance if someone creates a row manually. So, it's great to see that you are working on it, thanks!

**Comment by jeremyevans** [comment]
Can you try this patch?:

```ruby
diff --git a/lib/sequel/plugins/constraint_validations.rb b/lib/sequel/plugins/constraint_validations.rb
index c84179773..08b42fe4b 100644
--- a/lib/sequel/plugins/constraint_validations.rb
+++ b/lib/sequel/plugins/constraint_validations.rb
@@ -125,14 +125,15 @@ module Sequel
             ds = @dataset.with_quote_identifiers(false)
             table_name = ds.literal(ds.first_source_table)
             reflections = {}
-            @constraint_validations = (Sequel.synchronize{hash[table_name]} || []).map{|r| constraint_validation_array(r, reflections)}
+            allow_missing_columns = db_schema.select{|col, sch| sch[:allow_null] == false && nil != sch[:default]}.map(&:first)
+            @constraint_validations = (Sequel.synchronize{hash[table_name]} || []).map{|r| constraint_validation_array(r, reflections, allow_missing_columns)}
             @constraint_validation_reflections = reflections
           end
         end
 
         # Given a specific database constraint validation metadata row hash, transform
         # it in an validation method call array suitable for splatting to send.
-        def constraint_validation_array(r, reflections)
+        def constraint_validation_array(r, reflections, allow_missing_columns=EMPTY_ARRAY)
           opts = {}
           opts[:message] = r[:message] if r[:message]
           opts[:allow_nil] = true if db.typecast_value(:boolean, r[:allow_nil])
@@ -191,11 +192,13 @@ module Sequel
             reflection_opts[:argument] = arg
           end 
 
-          a << column
-          unless opts.empty?
-            a << opts
+          opts[:from] = :values
+          if column.is_a?(Symbol) && allow_missing_columns.include?(column)
+            opts[:allow_missing] = true
           end
 
+          a << column << opts
+
           if column.is_a?(Array) && column.length == 1
             column = column.first
           end
```

This should fix the issue. I'll do some more testing with it tomorrow, but assuming no problems, I'll probably commit it.

**Reply by 3ynm** [reply]
Wow, that was quick. Yes, that works indeed.
